core/src/main/java/jenkins/security/DefaultConfidentialStore.java

```diff
@@ -51,18 +68,26 @@ public DefaultConfidentialStore() throws IOException, InterruptedException {
     }
 
     public DefaultConfidentialStore(File rootDir) throws IOException, InterruptedException {
+        this(rootDir, getMasterKeyFile(rootDir));
+    }
+
+    protected DefaultConfidentialStore(File rootDir, File keyFile) throws IOException, InterruptedException {
         this.rootDir = rootDir;
         if (rootDir.mkdirs()) {
             // protect this directory. but don't change the permission of the existing directory
             // in case the administrator changed this.
             new FilePath(rootDir).chmod(0700);
         }
 
-        TextFile masterSecret = new TextFile(new File(rootDir, "master.key"));
+        TextFile masterSecret = new TextFile(keyFile);
         if (!masterSecret.exists()) {
-            // we are only going to use small number of bits (since export control limits AES key length)
-            // but let's generate a long enough key anyway
-            masterSecret.write(Util.toHexString(randomBytes(128)));
+            if (SystemProperties.getBoolean(MASTER_KEY_READONLY_SYSTEM_PROPERTY_NAME)) {
+                throw new IOException(masterSecret + " does not exist and system property " + MASTER_KEY_READONLY_SYSTEM_PROPERTY_NAME + " is set. You must provide a valid master key file.");
+            } else {
+                // we are only going to use small number of bits (since export control limits AES key length)
```

Comment:
BTW I think this comment has been obsolete for years?

Reply:
Still https://github.com/jenkinsci/jenkins/blob/2c92eae462cd6963462fc983458dd91f72494049/core/src/main/java/hudson/Util.java#L704

Reply:
Since Java 9 not true? https://stackoverflow.com/a/46857694/4951015

Reply:
Indeed, though would take non-trivial changes to migrate to the new key length. Not in scope of my current change.